src/nm-connectivity.c

```diff
@@ -483,35 +622,30 @@ nm_connectivity_check_start (NMConnectivity *self,
 		cb_data->ifspec = g_strdup_printf ("if!%s", iface);
 
 #if WITH_CONCHECK
-	if (iface) {
-		CURL *ehandle;
-
-		if (   priv->enabled
-		    && (ehandle = curl_easy_init ())) {
-
-			cb_data->concheck.response = g_strdup (priv->response);
-			cb_data->concheck.curl_ehandle = ehandle;
-			cb_data->concheck.request_headers = curl_slist_append (NULL, "Connection: close");
-			curl_easy_setopt (ehandle, CURLOPT_URL, priv->uri);
-			curl_easy_setopt (ehandle, CURLOPT_WRITEFUNCTION, easy_write_cb);
-			curl_easy_setopt (ehandle, CURLOPT_WRITEDATA, cb_data);
-			curl_easy_setopt (ehandle, CURLOPT_HEADERFUNCTION, easy_header_cb);
-			curl_easy_setopt (ehandle, CURLOPT_HEADERDATA, cb_data);
-			curl_easy_setopt (ehandle, CURLOPT_PRIVATE, cb_data);
-			curl_easy_setopt (ehandle, CURLOPT_HTTPHEADER, cb_data->concheck.request_headers);
-			curl_easy_setopt (ehandle, CURLOPT_INTERFACE, cb_data->ifspec);
-			curl_multi_add_handle (priv->concheck.curl_mhandle, ehandle);
-
-			cb_data->timeout_id = g_timeout_add_seconds (20, _timeout_cb, cb_data);
-
-			_LOG2D ("start request to '%s'", priv->uri);
-			return cb_data;
-		}
+	if (iface && ifindex > 0 && priv->enabled && priv->host) {
+		cb_data->concheck.ifindex = ifindex;
+		cb_data->concheck.resolve_cancellable = g_cancellable_new ();
+
+		g_dbus_proxy_new_for_bus (G_BUS_TYPE_SYSTEM,
```

Comment:
usually, we don't create new D-Bus proxies for each request. Also, we usually watch `name-owner` property, to know whether name resolving is available or not.

Reply:
> And I believe it's the wrong thing to do. Due to the asynchronous nature of DBusProxy creation, we'd need to make NMConnectivityCheck it async initable or hold back all 

The connectivity check is async already and tracked via that a `NMConnectivityCheckHandle` instance. If you create one proxy for all requests, that means, you need queuing handles while creating the proxy. Sure, that's mildly complicated, but seems very doable. For example, `src/nm-firewall-manager.c` does exactly that, in a very similar manner.

Not sure what you mean with `async initiable`. `NMConnectivityCheckHandle` is not a `GObject` and wouldn't implement `GAsyncInitable`.

Reply:
Yeah, I think NMFirewallManager is pretty terrible. It needs a global/singleton to keep track of in-flight requests. That I'd prefer to avoid.

Reply:
It does not need a global/singleton at all, it needs a NMFirewallManager instance -- which, since we only create one of such instance, is indeed a singleton. But there is nothing in code preventing you to create arbitrary many NMFirewallManager instances, all making their own requests.

I also don't see where it differs from NMConnectivity (as it is already today). Granted, firewall-manager makes an async D-Bus request, while nm-connectivity makes and async wget (+ creating a D-Bus proxy asynchronously). Apart from that their asynchronous operations do, they work fundamentally similar. They are also both "singleton" instances, and keep both track of their in-flight async handles.

I'm curious why you think NMFirewallManager is terrible and how it could be better. Could you elaborate?

Reply:
> I'm curious why you think NMFirewallManager is terrible and how it could be better. Could you elaborate?

It needs a logic to keep track of whether the proxy construction of the proxy was successful and track of in-flight requests. That adds extra moving parts at questionable benefit.

Reply:
With this reply I don't understand what you exactly suggest to improve it. Seem it will stay "terrible" and I will continue adding such code.

The main point of components like `NMFirewallManager` is to provide a *simple*, *correct* and *efficient* API to the upper layers. I clam it succeeds at doing that with mere 600 LOC, which is in the realm of being understandable. That is different from it's callers (`NMManager`(*) and `NMDevice`) which are complex beyond words. Looking at `NMFirewallManager` and saying that it's too complicated misses an important point: `NMFirewallManager` takes over a reasonable(?) amount of complexity, to make `NMManager` and `NMDevice` simpler. 

As can be seen with:
```
  $ git grep -30 nm_firewall_manager_ -- ':(exclude)src/nm-firewall-manager.?'
```

(*) `NMManager` does not call firewall, but `NMPolicy` does. However, `NMPolicy` *is* essentially a part of `NMManager`. There is no clear separation of responsibilities or API between them and they are strongly intermingled.

Reply:
An implementation using a single proxy is probably a bit more efficient
(how much, I can't say) but for sure it adds more complexity. In my
opinion, creating a proxy for each request is good enough and keeps the
code simple.